File: tmp/leetcode75-c/29.c
#define MASK 0x1FFF

struct TreeNode *create_tree(int *preorder,  int *preorder_idx, int *inorder_map,
                            int left, int right) {
    if (left > right)
        return NULL;

    struct TreeNode *root = malloc(sizeof(struct TreeNode));
    root->val = preorder[*preorder_idx];
    *preorder_idx += 1;
    root->left = create_tree(preorder, preorder_idx, inorder_map, left, inorder_map[root->val & MASK]-1);
    root->right = create_tree(preorder, preorder_idx, inorder_map, inorder_map[root->val & MASK]+1, right);
    return root;
}

struct TreeNode* buildTree(int* preorder, int preorderSize, int* inorder, int inorderSize){
    int preorder_idx = 0;
    int inorder_map[MASK+1] = {0};
    for (int i = 0; i < inorderSize; ++i) {
        inorder_map[inorder[i] & MASK] = i;
    }

    return create_tree(preorder, &preorder_idx, inorder_map, 0, inorderSize-1);
}

#undef MASK

File: tmp/leetcode75/29.py
'''
OLD NOTES:
      3
     /
    2
   /
  1
Preorder = 321  Inorder = 123 => reverse (no right child)

  3
   \
    2
     \
      1
Preorder = 321 Inorder = 321 => same (no left child)

     4
    /
   2
  / \
 1   3
Preorder = 421 3 Inorder = 12 3 4  => single right child
421 => 12 (right subtree of 2) 4

     6
    /
   2
  / \
 1   3
    / \
   4   5
Preorder = 621 345  Inorder = 12 435 6 => subtree right child
621 => 12 (right sub tree) 6

   2
    \
     3
    / \
   4   5
Preorder = 2345 Inorder = 2435

         6
        /
       2
      / \
     1   3
    /   / \
   7   4   5
  / \
 8   9
Preorder = 621789345 Inorder = 879124356

Algorithm:
push preorder nodes on stack until reach 1st inorder
    pop, curr is left child
    if next same index is equal...  right child.
    if next is not reversed: => right subtree.
        add (recurse / repeat) right tree
'''
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        def create_tree(left, right):
            if left > right:
                return None

            nonlocal preorder_idx
            root = TreeNode(preorder[preorder_idx])
            preorder_idx += 1
            root.left = create_tree(left, inorder_idx[root.val]-1)
            root.right = create_tree(inorder_idx[root.val]+1, right)
            return root

        preorder_idx = 0
        inorder_idx = {n:i for i, n in enumerate(inorder)}
        return create_tree(0, len(inorder)-1)
